### ISSUE TITLE
[bitname/rabbitmq] RollingUpdate taking replication into account?

### ISSUE THREAD

**Issue by msvticket** [bitnami/charts#3931]
**Which chart**:
rabbitmq

**Question**
When a the rabbitmq statefulset is updated and the update strategy is rollingUpdate the nodes are deleted and created again  successively. My understanding is that replicated queues then need to be replicated to a node when it comes up. Is there a mechanism so that it has replicated all messages it should before the next node is deleted?

I think this can also be asked like this: Does `rabbitmq-diagnostics -q check_running` report the node as down as long as it is still replicating after being started?

If the answer is no there is a risk that an update leads to loosing data if there are a large amount of messaging in replicated queues in the cluster. This is something that happens when using the rabbitmq-ha chart and I would like to know if this problem has been addressed in the bitnami rabbitmq chart.

I'm sorry if there is another preferred channels for questions like this, but I couldn't find it.

**Comment by andresbono** [comment]
Hi @msvticket, thanks for opening this issue and raising this interesting topic. The livenessProbe is configured to execute the command that you mention: https://www.rabbitmq.com/rabbitmq-diagnostics.8.html#check_running

https://github.com/bitnami/charts/blob/3748bad96227b030f86f50f535c3ad71a092bb10/bitnami/rabbitmq/templates/statefulset.yaml#L202-L207

However, I'm not really sure if if that command takes into account the replication state or not. Is there any other check that you can suggest? We will be happy to continue improving this chart.

**Comment by msvticket** [comment]
That's what I hoped you would know... ;) Looking in the documentation I don't find that it does or that any similar command exists. What exists are checks to see if a node can be shutdown without problem for replication: `rabbitmq-diagnostics -q check_if_node_is_mirror_sync_critical` and `rabbitmq-diagnostics -q check_if_node_is_quorum_critical`

So what could be done to lessen the risk I'm describing would be to change the pre stop hook to not run "rabbitmqctl stop_app" until the checks mentioned above return OK. Such a pre stop script would need to take into account that these checks also fail if the node is down already. The terminationGracePeriodSeconds setting would need to be increased but also the script should take that setting into account and give up the checks even if they fail if the grace period is about to end.

https://github.com/bitnami/charts/blob/master/bitnami/rabbitmq/templates/statefulset.yaml#L253

**Comment by andresbono** [comment]
@msvticket, the approach sounds good to me, it seems safer than just a simple `rabbitmqctl stop_app`. Would you like to contribute with a PR addressing the changes?

**Comment by msvticket** [comment]
Yeah. I think I can get time to make a PR within a couple of weeks. Shall I reword this story in the form of a feature request instead?

**Comment by andresbono** [comment]
That would be great, @msvticket!! ❤️  About rewording this thread, I think it's fine to just fill the [PR template](https://github.com/bitnami/charts/blob/master/.github/PULL_REQUEST_TEMPLATE.md) to describe the changes of the new PR and add the link to this issue.

**Comment by msvticket** [comment]
I have started working on this, but it will take a bit longer than expected since I first need to fix  #3675 to be able to use the chart at all.

**Comment by andresbono** [comment]
Thanks for the update, @msvticket 🎉 .

**Comment by carrodher** [comment]
Thanks for the PR in the container repository. It would be nice to have a PR in the chart using the new scripts, in that way we can handle the release of the container plus the chart together.
I just set the "on-hold" label to prevent the stale bot from closing this issue

**Comment by msvticket** [comment]
This is not really fixed since bitnami/bitnami-docker-rabbitmq#142 hasn't been merged...

**Comment by mltsy** [comment]
Ah - wonderful! 🎉 

There was a recent related PR adjusting the livenessProbe: https://github.com/bitnami/charts/pull/4598

I was just discussing there whether we should have an option to wait for syncing to occur there, but I think *if* the fix here postpones the termination of a pod during an upgrade until it is not quorum-critical or mirror-critical, that's a much better solution, and we probably don't even need the startup to wait for sync to occur before starting the next node, right?  (So `rabbitmq-diagnostics -q ping` is fine for a livenessProbe - the livenessProbe should not have anything to do with syncing ideally, but previously there was just no other check, so this is great!)

**Comment by andresbono** [comment]
@mltsy, once https://github.com/bitnami/bitnami-docker-rabbitmq/pull/142 (or equivalent PR) is merged and this new functionality is in place, would you like to help test the scenarios that you are proposing? Then, we can do follow-up improvements based on that testing if needed.

**Comment by mltsy** [comment]
If I can find time for the testing, I'd like to.  But if not, as I [commented here](https://github.com/bitnami/charts/pull/4470#issuecomment-742600650) I think perhaps all we need is some documentation about how it's expected to work!  I think the new solution might actually be sufficient, as long as it's documented.  (And at least once it's documented, people will have a reference point to write up issues against if it doesn't work as described/expected)

**Comment by mltsy** [comment]
(Which I'm hoping I can find time for this month!)

**Comment by carrodher** [comment]
Thanks, yes, I agree that the best solution apart from having the technical part done is also to add the proper documentation. Let's continue iterating over this in order to find a good approach including docs as part of this work
